pytorch_lightning/trainer/supporters.py

```diff
@@ -371,6 +370,64 @@ def __init__(self, loaders: Any, mode: str = 'min_size'):
         if self.mode == 'max_size_cycle':
             self._wrap_loaders_max_size_cycle()
 
+        self.loaders_iter_state_dict = None
+
+    def state_dict(self, num_batches_processed: int):
+        """
+        The state dict includes all states from wrapped dataloaders and their samplers through the
+        ``CaptureIterableDataset`` and fast-forward samplers.
+
+        Args:
+            num_batches_processed: The number of batches processed so far, needed because the individual dataloaders
+                may have already prefetched more batches by the time a state dict is requrested.
+        """
+        if not fault_tolerant_enabled():
+            return DataLoaderDict()
+
+        def state_dict_fn(dataloader: DataLoader, iterator: Iterator) -> Dict:
+            # find next worker if multiple workers were used
+            state = find_current_worker(iterator)
+            if isinstance(dataloader.dataset, CaptureIterableDataset):
+                # the sampler state dict are extracted in ``CombinedLoaderIterator``
+                if iterator is not None and getattr(iterator, "_sampler_state_dict", None) is not None:
```

Comment:
why the `iterator is not None` check? It is not marked as `Optional` in the signature

Reply:
I think this is because we are not always dealing with DataLoaders but users can also use custom iterators for which we then cannot get a statedict (covered by the defaultvalue of `None`). So we cannot update it in that case.

cc @tchaton  to confirm

Reply:
In that case, the typing should be updated.

And maybe the `None` check should be done right at the start of the function

Reply:
The iterator is potentially None if `iter` has never been called on CombinedDataLoader.